Fix Pydantic Models to support V1 and V2 of Pydantic and fix deprecation warnings

njbrake opened the pull request:
### Related Issues/PRs

I see https://github.com/mlflow/mlflow/pull/13023 was also started to try to fix it, but it appears to have lost momentum.

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> [#12654]

### What changes are proposed in this pull request?

NFC: update the pydantic model names to avoid getting warnings with newer pydantic versions

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [ ] API references
  - [x] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

njbrake committed:
updating pydantic models

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
sign off

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
Simplified validator toggle for backwards compatibility

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
don't touch infra code

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
making it work as a decorator

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
making it work as a decorator

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
making it work as a decorator

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
sign off

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
patch decorator

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
patch up

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
runtime checkable

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
patch up

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
searching for a nice solution

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
searching for a nice solution

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
searching for a nice solution

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake commented:
Sorry for all the commits 🙈 , this will certainly want to be a squash merge. Anyways, I think I ended up with a nice solution that will allow us to support both Pydantic V1 and V2 while keeping the code reasonably clean. Let me know what you think and feel free to merge in if you agree.

bonastreyair commented on mlflow/utils/pydantic_utils.py:
```diff
@@ -7,6 +7,17 @@
 IS_PYDANTIC_V2_OR_NEWER = Version(pydantic.VERSION).major >= 2
 
 
+def validator(field:str, pre: bool=True):
+    def decorator(func: Callable) -> Callable:
+        if IS_PYDANTIC_V2_OR_NEWER:
+            from pydantic import field_validator as pydantic_validator 
+            return pydantic_validator(field, mode="before" if pre else "after")(func)
+        else:
+            from pydantic import validator as pydantic_validator
+            return pydantic_validator(field, pre=True)(func) # 
+
+    return decorator
```
what about the model validator?

```python
def pydantic_model_validator(mode: str):
    def decorator(func: Callable) -> Callable:
        if IS_PYDANTIC_V2:
            from pydantic import model_validator

            return model_validator(mode=mode)(func)
        else:
            from pydantic import root_validator

            pre = mode == "pre"
            return root_validator(pre=pre)(func)
```

bonastreyair replied:
basically to also fix 

```python
    if IS_PYDANTIC_V2:
        from pydantic import model_validator as _model_validator

        @_model_validator(mode="before")
        def validate_field_compatibility(cls, info: Dict[str, Any]):
            return cls._validate_field_compatibility(info)

    else:
        from pydantic import root_validator as _root_validator

        @_root_validator(pre=False)
        def validate_field_compatibility(cls, config: Dict[str, Any]):
            return cls._validate_field_compatibility(config)
```

from mlflow.gateway.config.py

bonastreyair commented on docs/docs/llms/deployments/index.mdx:
```diff
@@ -1469,7 +1469,7 @@ and a config class that inherits from `mlflow.gateway.base_models.ConfigModel`.
 from typing import AsyncIterable
 
-from pydantic import validator
+from mlflow.utils import validator
```
isn't it clearer to import from `mlflow.utils.pydantic`? or `mlflow.utils.pydantic_utils`?

bonastreyair commented on mlflow/utils/pydantic_utils.py:
```diff
@@ -7,6 +7,17 @@
 IS_PYDANTIC_V2_OR_NEWER = Version(pydantic.VERSION).major >= 2
 
 
+def validator(field:str, pre: bool=True):
```
I would recommend the validator to follow the "new kwargs from v2, using `mode=`" and adapt it for V1 (and not the other way around), so that whenever V1 gets removed it is just removing that adaptation rather than modifying every validator.

right now it seems a working patch, but it could become a easier journey for the future

njbrake committed:
added model validator

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake commented:
Thanks, @bonastreyair for the helpful comments! I think I addressed all of them, and hopefully also fixed an issue that was causing some unit tests to fail. I was initially confused why the below lines existed in the current codebase:

```
    if IS_PYDANTIC_V2_OR_NEWER:

        @field_validator("config", mode="before")
        def validate_config(cls, info, values):
            return cls._validate_config(info, values)
    else:

        @field_validator("config", mode="before")
        def validate_config(cls, config, values):
            return cls._validate_config(config, values)
            ```
I think those differing param names (info + values vs config + values) are used for some sort of check that is inferring a signature, so I need to keep that logic in the gateway.config.py file after all.

njbrake committed:
Feature parity with previous implementation

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
fixing model_validator code

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
fixing model_validator code

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
fixing model_validator code

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
fixing model_validator code

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
fixing model_validator code

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

bonastreyair commented on mlflow/gateway/config.py:
```diff
@@ -159,16 +159,13 @@ def _validate_field_compatibility(cls, info: dict[str, Any]):
 
-        @_model_validator(mode="before")
+        @model_validator(mode="after")
         def validate_field_compatibility(cls, info: dict[str, Any]):
             return cls._validate_field_compatibility(info)
-
     else:
-        from pydantic import root_validator as _root_validator
 
-        @_root_validator(pre=False)
+        @model_validator(mode="after")
         def validate_field_compatibility(cls, config: dict[str, Any]):
             return cls._validate_field_compatibility(config)
```
I think this if can now be deleted from here no? I mean it is exactly the same code

bonastreyair commented on mlflow/types/agent.py:
```diff
@@ -18,12 +18,7 @@
-    from pydantic import model_validator
-else:
-    from pydantic import root_validator
+from mlflow.utils import IS_PYDANTIC_V2_OR_NEWER, model_validator
```
```suggestion
from mlflow.uti.pydantic_utils import IS_PYDANTIC_V2_OR_NEWER, model_validator
```

bonastreyair commented on mlflow/gateway/providers/mlflow.py:
```diff
@@ -1,19 +1,20 @@
 from mlflow.gateway.providers.base import BaseProvider
 from mlflow.gateway.providers.utils import send_request
 from mlflow.gateway.schemas import chat, completions, embeddings
+from mlflow.utils import field_validator
```
```suggestion
from mlflow.utils.pydantic_utils import field_validator
```

bonastreyair commented on examples/gateway/plugin/my-llm/my_llm/config.py:
```diff
@@ -1,14 +1,13 @@
     my_llm_api_key: str
 
-    @validator("my_llm_api_key", pre=True)
+    @field_validator("my_llm_api_key", pre=True)
```
```suggestion
    @field_validator("my_llm_api_key", mode="before")
```

bonastreyair commented on examples/gateway/plugin/my-llm/my_llm/config.py:
```diff
@@ -1,14 +1,13 @@
-from pydantic import validator
-
 from mlflow.gateway.base_models import ConfigModel
+from mlflow.utils import field_validator
```
```suggestion
from mlflow.utils.pydantic_utils import field_validator
```
better explicit than implicit

bonastreyair commented on docs/docs/llms/deployments/index.mdx:
```diff
@@ -1469,7 +1469,7 @@ and a config class that inherits from `mlflow.gateway.base_models.ConfigModel`.
 from typing import AsyncIterable
 
-from pydantic import validator
+from mlflow.utils import field_validator
```
```suggestion
from mlflow.utils.pydantic_utils import field_validator
```

bonastreyair commented on mlflow/gateway/config.py:
```diff
@@ -26,7 +26,7 @@
     is_valid_mosiacml_chat_model,
 )
-from mlflow.utils import IS_PYDANTIC_V2_OR_NEWER
+from mlflow.utils import IS_PYDANTIC_V2_OR_NEWER, field_validator, model_validator
```
```suggestion
from mlflow.utils.pydantic_utils import IS_PYDANTIC_V2_OR_NEWER, field_validator, model_validator
```

bonastreyair commented on mlflow/gateway/config.py:
```diff
@@ -298,26 +295,31 @@ def validate_provider(cls, value):
-            return config_type(**info)
+        # For Pydantic v2: 'context' is a ValidationInfo object with a 'data' attribute.
+        # For Pydantic v1: 'context' is dict-like 'values'.
+        if IS_PYDANTIC_V2_OR_NEWER:
+            provider = context.data.get("provider")
+        else:
+            provider = context.get("provider") if context else None
```
I would also move this into the pydantic_utils, so that the day we want to finally remove pydantic v1, we would just need to delete the pydantic_utils file

bonastreyair commented on mlflow/gateway/config.py:
```diff
@@ -372,10 +374,14 @@ def validate_model(cls, model):
     def validate_route_type_and_model_name(cls, values):
-        route_type = values.get("route_type")
-        model = values.get("model")
+        if IS_PYDANTIC_V2_OR_NEWER:
+            route_type = values.route_type
+            model = values.model
+        else:
+            route_type = values.get("route_type")
+            model = values.get("model")
```
same here, logic to be moved to pydantic_utils

bonastreyair commented on mlflow/types/agent.py:
```diff
@@ -55,7 +50,7 @@ class ChatAgentMessage(BaseModel):
     if IS_PYDANTIC_V2_OR_NEWER:
 
-        @model_validator(mode="after")
+        @model_validator
```
this will fail when used, since the model_validator does not contain a mode, but then the decorator actually passes a mode... so decorator will not be able to instanciate the function

njbrake replied:
Thanks, good catch! I'm having a hard time checking my work, because the unit tests all pass locally but I'm not confident that the tests I'm running are really checking everything. Sorry for all the commits 🙈

bonastreyair commented on mlflow/types/agent.py:
```diff
@@ -55,7 +50,7 @@ class ChatAgentMessage(BaseModel):
 
     if IS_PYDANTIC_V2_OR_NEWER:
```
can this logic be translated into a single def check_content_and_tool_calls

and move the logic to either inside the function or into pydantic_utils files?

njbrake replied:
Refactored to consolidate the function, but I lean towards keeping the logic here instead of moving it to pydantic_utils, if that's ok.

bonastreyair commented on mlflow/gateway/config.py:
```diff
@@ -159,16 +159,13 @@ def _validate_field_compatibility(cls, info: dict[str, Any]):
-        from pydantic import model_validator as _model_validator
 
-        @_model_validator(mode="before")
+        @model_validator(mode="after")
```
are you sure about this change from "before" to "after?

bonastreyair commented on mlflow/gateway/config.py:
```diff
@@ -372,10 +374,14 @@ def validate_model(cls, model):
                 )
         return model
 
-    @root_validator(skip_on_failure=True)
```
what about this logic? are we losing it?

njbrake replied:
skip_on_failure is no longer available in Pydantic V2, unfortunately.

bonastreyair replied:
Yeah but for pydantic v1 it still needs to work since it is supported

njbrake replied:
Good point. Updated to address that comment 👍

njbrake committed:
Apply suggestions from code review

Co-authored-by: Yair Bonastre <37800834+bonastreyair@users.noreply.github.com>
Signed-off-by: Nathan Brake <33383515+njbrake@users.noreply.github.com>

njbrake committed:
Updates from PR Review

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
Updates from PR Review

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
format

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake committed:
Updates from PR Review

Signed-off-by: Nathan Brake <nathan@mozilla.ai>

njbrake commented:
Hi @B-Step62 ! Are you the correct Mlflow maintainer to contact about this PR? @bonastreyair and I have done some work to get this PR pulled together and I think it is ready for you (and at least ready to be subjected to the full test suite). Thanks!

B-Step62 commented:
Thank you for much for the effort, @njbrake! Started the full CI, will take a path shortly as well.

B-Step62 reviewed (approved):
Works like a charm, thank you so much for solving the large devex pain!!

The spacy 3.3 test can be ignored, we are dropping the support for that version in https://github.com/mlflow/mlflow/pull/14594
